burgerNode.js

```diff
@@ -29,8 +29,108 @@ class BurgerNode {
+
+      for (var i = 1; i < newChain.blocks.length; i++) {
+        const newBlock = newChain.blocks[i]
+        // newBlock.prevBlockhash = "somestring"; // just for testing purposes!!!
+        const areBlockKeysAndValuesValid = BurgerBlock.validateBlock(newBlock); // just validating keys/values
+
+        // re-calculate block data hash
+        const compareBlock = new BurgerBlock(
+          newBlock.index,
+          newBlock.transactions,
+          newBlock.difficulty,
+          newBlock.prevBlockhash,
+          newBlock.minedBy
+        );
+        const doBlockDataHashesMatch = newBlock.blockDataHash === compareBlock.blockDataHash;
+
+        // re-calculate block hash
+        const doBlockHashesMatch = this.chain.isBlockValid(newBlock);
+
+        // corroborate hash's difficulty
+        const currentDifficulty = newBlock.difficulty;
+        const isDifficultyValid = newBlock.blockHash.substring(0, currentDifficulty) === Array(currentDifficulty + 1).join('0');
+
+        // corroborate the previousHash value of the previous block
+        const previousBlock = newChain.blocks[i - 1];
+        const isPrevBlockHashValid = newBlock.prevBlockhash === previousBlock.blockHash;
+
+        if (!areBlockKeysAndValuesValid
+          || !doBlockDataHashesMatch
+          || !doBlockHashesMatch
+          || !isDifficultyValid
+          || !isPrevBlockHashValid
+        ) {
+          areBlocksValid = false;
+          break;
+        }
+
+        let areTransactionsValid = this.validateTransactionsOfBlock(newBlock);
+
+        if (!areTransactionsValid) {
+          areBlocksValid = false;
+          break;
+        }
+      }
+
+      return areBlocksValid;
+    }
+
+    validateGenesisBlock(newChain) {
+      // [Anar] maybe this logic should go in the chain?
+      const genesisBlockHeld = this.chain.blocks[0].transactions[0];
+      const genesisBlockHeldKeys = Object.keys(genesisBlockHeld);
+      const genesisBlockHeldValues = Object.values(genesisBlockHeld);
+
+      const receivedGenesisBlock = newChain.blocks[0].transactions[0];
+      const receivedGenesisBlockKeys = Object.keys(receivedGenesisBlock);
+      const receivedGenesisBlockValues = Object.values(receivedGenesisBlock);
+
+      const areKeysEqual = JSON.stringify(genesisBlockHeldKeys) === JSON.stringify(receivedGenesisBlockKeys);
+      const areValuesEqual = JSON.stringify(genesisBlockHeldValues) === JSON.stringify(receivedGenesisBlockValues);
+
+      return areKeysEqual && areValuesEqual;
+    }
+
+    validateChain(newChain) {
+      const isGenesisBlockValid = this.validateGenesisBlock(newChain);
+      const areBlocksValid = this.validateBlocks(newChain);
+
+      return isGenesisBlockValid
+          && areBlocksValid;
+    }
+
     replaceChain(newChain) {
+      const isChainValid = this.validateChain(newChain);
+      const hasMoreCumulativeDifficulty = newChain.calculateCumulativeDifficulty() > this.chain.calculateCumulativeDifficulty();
```

Comment:
`newChain.calculateCumulativeDifficulty()` here will be undefined.

`newChain` is transmitted as a plain JSON object. Needs to transform to a `BurgerBlockchain` instance first before the method is accessed.